cmd/skaffold/app/cmd/parse_config.go

```diff
@@ -108,11 +108,18 @@ func getConfigs(cfgOpts configOpts, opts config.SkaffoldOptions, r *record) ([]*
 func processEachConfig(config *latest.SkaffoldConfig, cfgOpts configOpts, opts config.SkaffoldOptions, r *record, index int) ([]*latest.SkaffoldConfig, error) {
 	// check that the same config name isn't repeated in multiple files.
 	if config.Metadata.Name != "" {
-		prevConfig, found := r.configNameToFile[config.Metadata.Name]
-		if found && prevConfig != cfgOpts.file {
-			return nil, fmt.Errorf("skaffold config named %q found in multiple files: %q and %q", config.Metadata.Name, prevConfig, cfgOpts.file)
+		prev, found := r.configNameToFile[config.Metadata.Name]
+		if found {
+			sl := strings.SplitN(prev, "@", 2) // map value is always formatted as `file_name@config_index`
+			prevConfig, prevIndex := sl[0], sl[1]
+			if prevConfig != cfgOpts.file {
+				return nil, fmt.Errorf("skaffold config named %q found in multiple files: %q and %q", config.Metadata.Name, prevConfig, cfgOpts.file)
+			}
+			if prevIndex != fmt.Sprint(index) {
+				return nil, fmt.Errorf("multiple skaffold configs named %q found in file %q", config.Metadata.Name, cfgOpts.file)
+			}
```

Comment:
Can duplicate references not be handled by `ParseConfigAndUpgrade()`?  So all we'd need is a set of seen config names?

Reply:
It can't exactly be inside `ParseConfigAndUpgrade()` since that works with `VersionedConfig` which doesn't have any method like `GetName`. To make that change I'd have to modify every version of the skaffold config to satisfy that interface. But you're right in that I can do this check upfront. Moved.